app/Singletons/Mods.php

```diff
@@ -135,19 +135,31 @@ public function idsToBitset($ids): int
         );
     }
 
-    public function assertValidExclusivity(int $rulesetId, array $requiredIds, array $allowedIds): bool
+    public function assertValidExclusivity(int $rulesetId, array $modAcronyms): bool
     {
+        while (($modAcronym = array_pop($modAcronyms)) !== null) {
+            $mod = $this->mods[$rulesetId][$modAcronym];
+            $incompatibleIds = $mod['IncompatibleMods'];
+
+            $invalidIds = $incompatibleIds->intersect(new Set($modAcronyms));
+            if ($invalidIds->count() > 0) {
+                throw new InvariantException("incompatible mods: {$modAcronym}, {$invalidIds->join(', ')}");
+            }
+        }
+
+        return true;
+    }
+
+    public function assertValidMultiplayerExclusivity(int $rulesetId, array $requiredIds, array $allowedIds): bool
+    {
+        $this->assertValidExclusivity($rulesetId, $requiredIds);
+
         $disallowedIds = new Set();
 
         while (($requiredId = array_pop($requiredIds)) !== null) {
             $mod = $this->mods[$rulesetId][$requiredId];
             $incompatibleIds = $mod['IncompatibleMods'];
```

Comment:
as mentioned in other comment,
```suggestion
        foreach ($requiredIds as $requiredId) {
            $incompatibleIds = $this->mods[$rulesetId][$requiredId]['IncompatibleMods'];
```

Reply:
I can't see a reason why both loops can't use `foreach` so I did both.

Reply:
hmm, I wonder if this one can just gather all `$incompatibleIds` first and then compare with `$modAcronyms`. The `$modAcronym` info will be gone but it shouldn't happen in the first place if using client...?

I think the original idea was the modAcronyms array gets smaller as it goes

Reply:
Applied in https://github.com/ppy/osu-web/pull/11875/commits/c81810e1701be5df2e507b2fcca7babd3a56372a, I think? I don't see any reason why it wouldn't work.